### PR TITLE
Update rand() documentation to be more modern and clear

### PR DESCRIPTION
**Pull request by scottchiefbaker**
Learning more about how Perl implements random numbers lead me to clean up and update the documentation.

### CODE REVIEW

File: pod/perlfunc.pod
Comment (bram-perl) [inline pod/perlfunc.pod]: Extra `>` at the end of the sentence?

I would also add something like: 'see documentation of rand for alternatives`

File: pod/perlfunc.pod
Comment (bram-perl) [inline pod/perlfunc.pod]: I would put this change in a separate commit.

The rest of the changes are mostly formulating it slightly differently/reordering the existing text but this  appears to be completely new.

Also: if you want to make this particular claim then *I* would also like to know on what it is based.
This is making it very scary and will scare users and personally *I* do not know if that is warranted.

File: pod/perlfunc.pod
Comment (bram-perl) [inline pod/perlfunc.pod]: Could be useful to make it more explicit,

i.e. something like:
```
    my $num1 = rand();        # Random float between 0 and 1  (0 <= rand() < 1)
    my $num2 = rand(10);      # Random float between 0 and 10  (0 <= rand(10) < 10)
    my $num3 = int(rand(10)); # Random integer between 0 and 10 (0 <= int(rand(10)) < 10) 
```

(When using `between x and y` I always have some doubt if y is included or not. It may or may not be so I would make it very explicit)

File: pod/perlfunc.pod
Comment (bram-perl) [inline pod/perlfunc.pod]: (I'm aware the message was copied from the original)

It's not **my** `rand` function. It's perl `rand` function.
--> I would phrase it as: `If the rand function consistently ....`

(Also one has to wonder what "too large or too small" really mean. I guess I should build a perl with a wrong number of RANDBITS to find out..)

Comment (scottchiefbaker) [reply pod/perlfunc.pod]: I was tempted to remove this line altogether. There isn't much info here and this is really a bug and maybe not worth noting here?

File: pod/perlfunc.pod
Comment (Grinnz) [reply pod/perlfunc.pod]: I think the current note later on that it is not cryptographically secure is sufficient, and perhaps this note could accompany that as further information, but I agree that it doesn't seem useful to proscribe what sort of uses a non-cryptographically-secure PRNG is acceptable for. * without reliable references or evidence.

File: pod/perlfunc.pod
Comment (bram-perl) [reply pod/perlfunc.pod]: The note appears to go back to perl v5.0.0.. so it's been there for a while..

I'm a bit split on removing it..

**Assuming** this is still a problem that can happen:
- For 99.999% of the people reading the docs it will be of no value
- For the person that is running into this issue having the information however is **invaluable** because there will not be an easy way for anyone to figure out what the underlying problem is

I think I would only be in favor of removing it if:
- the problem can no longer occur
- the problem can occur but there is a very clear warning/error/... that it is happening (for example a warning when `rand()` is being used.)

File: pod/perlfunc.pod
Comment (scottchiefbaker) [reply pod/perlfunc.pod]: That trailing `>` is in the original comment. I believe that's the closing tag for the bold on the text. But yes I agree about adding a link to "see rand"

File: pod/perlfunc.pod
Comment (scottchiefbaker) [reply pod/perlfunc.pod]: Personally I think this is too verbose. It should be clear from the sentence before this what the exact range is.

Comment (bram-perl) [reply pod/perlfunc.pod]: In that case the comments can be completely dropped since that should also be clear from the previous sentence. :-)


Either they read the previous sentences (and then they know the exact range and there is no need for the comment) or they didn't and in that case it is ambiguous.

File: pod/perlfunc.pod
Comment (scottchiefbaker) [reply pod/perlfunc.pod]: Heh... I asked ChatGPT if `drand48()` was any good and this is word-for-word what it said :)

**Comment by scottchiefbaker** [comment]
Any more advice on fixes/updates on wording so this can get merged?

File: pod/perlfunc.pod
Comment (guest20) [reply pod/perlfunc.pod]: > `your C<rand>`

This refers to the `rand` in the libs/OS you picked. This type of language is used to couch all sorts of porting issues.

**Comment by demerphq** [comment]
On Tue, 12 Nov 2024, 14:20 guest20, ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In pod/perlfunc.pod
> <https://github.com/Perl/perl5/pull/22727#discussion_r1838089520>:
>
> > -(Note: If your rand function consistently returns numbers that are too
> +=item *
> +
> +Currently EXPR with the value C<0> is
> +also special-cased as C<1>. This was undocumented before Perl 5.8.0
> +and is subject to change in future versions of Perl.
> +
> +=item *
> +
> +As of Perl v5.20.0 C<rand> uses the C<drand48()> PRNG to generate random numbers.
> +This should be acceptable for educational or hobby projects, basic simulations,
> +or simple random sampling.
> +
> +=item *
> +
> +If your C<rand> function consistently returns numbers that are too
>
> your C<rand>
>
> This refers to the rand in the libs/OS you picked. This type of language
> is used to couch all sorts of porting issues.
>

Any such language should be removed, we added the hand rolled drand48()
specifically to insulate ourselves from OS dependencies.  I don't remember
if we support a build option to use the system rand/srand (but probably we
do), if we do then we can keep these comments but caveat them
appropriately; a normal build will use the same (bad) rng everywhere.

Yves

>

**Comment by demerphq** [comment]
On Tue, 12 Nov 2024, 06:44 Scott Baker, ***@***.***> wrote:

> Any more advice on fixes/updates on wording so this can get merged?
>

FWIW i don't think we should say that our rand is suitable for any specific
purpose. I think we should say that it is not suitable for any
mathematically rigorous purpose and leave it to the reader to make their
own decisions.  Perhaps suggesting some alternatives like MT or WELL
variants. WELL-512 is probably one of the better choices these days.

Linear congruent random number generators as a whole have a variety of
issues, and the one we provide was modern when 16 bit integers were modern
and has less state (48 bits of course) than an integer in a standard
build.  To make things worse iirc we don't even dispose of the high bits in
the state, which are substantially less random than the low bits.

These days the state of art of non cryptographic generators with limited
state is much advanced and people really shouldn't use rand() for anything
serious at all. I really wish we bundled a better RNG.

Yves

>

**Comment by scottchiefbaker** [comment]
@demerphq I updated the two sections you referenced.

File: pod/perlfunc.pod
Comment (scottchiefbaker) [reply pod/perlfunc.pod]: I took out the wording that ChatGPT used

**Comment by demerphq** [comment]
On Thu, 14 Nov 2024, 03:32 Scott Baker, ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In pod/perlfunc.pod
> <https://github.com/Perl/perl5/pull/22727#discussion_r1841479214>:
>
> > +As of Perl v5.20.0 C<rand> uses the C<drand48()> PRNG to generate random numbers.
> +This should be acceptable for educational or hobby projects, basic simulations,
> +or simple random sampling.
> +
>
> I took out the wording that ChatGPT used
>
Thanks.

Unfortinately it still says: "As a PRNG drand48() should be sufficient for
most non-cryptographic needs."

I don't think this is correct, an LCG PRNG is not suitable for many tasks,
for instance if you use an LCG to choose coordinates you will find that the
coordinates chosen all lie on a predictable hyper-lattice. (Wikipedia has a
nice animated gif showing this IIRC)

IMO Drand48 is basically only suitable for simple sampling tasks which do
not require any mathematical rigor.

Cheers
Yves

File: pod/perlfunc.pod
Comment (leonerd) [reply pod/perlfunc.pod]: I'd say:

```perl
    my $num1 = rand();        # Random float at least 0 and below 1
    my $num2 = rand(10);      # Random float at least 0 and below 10
    my $num3 = int(rand(10)); # Random integer at least 0 and below 10
```

thus emphasising that it avoids the upper bound here. "between" is too ambiguous a word.

File: pod/perlfunc.pod
Comment (leonerd) [inline pod/perlfunc.pod]: The word "also" here seems to be left over from the previous version. I'd remove it, and just say "0 is special-cased to mean 1".
Also this paragraph needs reflowing.

Comment (scottchiefbaker) [reply pod/perlfunc.pod]: Good catch... this is corrected.

File: pod/perlfunc.pod
Comment (khwilliamson) [reply pod/perlfunc.pod]: `least .. below` lgtm

Comment (scottchiefbaker) [reply pod/perlfunc.pod]: That's fair... PR updated.

File: pod/perlfunc.pod
Comment (leonerd) [inline pod/perlfunc.pod]: Lowercase "automatically"

**Review by leonerd — APPROVED** [review]
Needs one letter case adjustment but aside from that, LGTM

**Comment by scottchiefbaker** [comment]
@leonerd good catch... docs updated

**Comment by khwilliamson** [comment]
@scottchiefbaker These commits should be squashed before we merge

File: pod/perlfunc.pod
Comment (khwilliamson) [inline pod/perlfunc.pod]: Somehow I think we use B<> instead of italics for these situations

File: pod/perlfunc.pod
Comment (khwilliamson) [inline pod/perlfunc.pod:6572]: I'm unsure B<> is appropriate here

File: pod/perlfunc.pod
Comment (khwilliamson) [inline pod/perlfunc.pod]: PRNG is that abbreviation really obvious?

File: pod/perlfunc.pod
Comment (scottchiefbaker) [reply pod/perlfunc.pod:6572]: We boldify the "Security:" section that's below this. I think making "Notes:" match is best for consistency.

File: pod/perlfunc.pod
Comment (scottchiefbaker) [reply pod/perlfunc.pod]: Are you suggesting that we spell out "pseudorandom number generator" instead of PRNG?

File: pod/perlfunc.pod
Comment (scottchiefbaker) [reply pod/perlfunc.pod]: Fair enough. I landed this change.

File: pod/perlfunc.pod
Comment (scottchiefbaker) [reply pod/perlfunc.pod]: Done

**Pull request merged**
